File: stdlib/js/hull/middleware/tests/test_inbox.js
// test_inbox.js — Tests for hull:middleware:inbox
//
// Requires db, time globals (run via hull test harness).

import { inbox } from "hull:middleware:inbox";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

// ── init ─────────────────────────────────────────────────────────────

test("init creates inbox table", () => {
    inbox.init({ ttl: 60 });
});

// ── isDuplicate ─────────────────────────────────────────────────────

test("new message is not a duplicate", () => {
    assertEq(inbox.isDuplicate("js-msg-new-1", "test"), false);
});

test("null messageId returns false", () => {
    assertEq(inbox.isDuplicate(null, "test"), false);
});

test("empty messageId returns false", () => {
    assertEq(inbox.isDuplicate("", "test"), false);
});

// ── markProcessed ───────────────────────────────────────────────────

test("markProcessed then isDuplicate returns true", () => {
    inbox.markProcessed("js-msg-mark-1", "test");
    assertEq(inbox.isDuplicate("js-msg-mark-1", "test"), true);
});

test("different source is not duplicate", () => {
    inbox.markProcessed("js-msg-src-1", "source-A");
    assertEq(inbox.isDuplicate("js-msg-src-1", "source-B"), false);
});

// ── checkAndMark ─────────────────────────────────────────────────────

test("checkAndMark returns false for new message", () => {
    const dup = inbox.checkAndMark("js-msg-cam-1", "test");
    assertEq(dup, false, "first call should not be duplicate");
});

test("checkAndMark returns true for second call", () => {
    inbox.checkAndMark("js-msg-cam-2", "test");
    const dup = inbox.checkAndMark("js-msg-cam-2", "test");
    assertEq(dup, true, "second call should be duplicate");
});

// ── default source ───────────────────────────────────────────────────

test("default source works", () => {
    inbox.markProcessed("js-msg-default-1");
    assertEq(inbox.isDuplicate("js-msg-default-1"), false);
    assertEq(inbox.isDuplicate("js-msg-default-1", "default"), true);
});

// ── cleanup ─────────────────────────────────────────────────────────

test("cleanup returns count", () => {
    const count = inbox.cleanup();
    assertEq(typeof count, "number");
});

export default { pass, fail };

File: stdlib/lua/hull/middleware/tests/test_inbox.lua
-- test_inbox.lua — Tests for hull.middleware.inbox
--
-- Requires db, time globals (run via hull test harness).

local inbox = require('hull.middleware.inbox')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- ── init ─────────────────────────────────────────────────────────────

test("init creates inbox table", function()
    inbox.init({ ttl = 60 })
end)

-- ── is_duplicate ─────────────────────────────────────────────────────

test("new message is not a duplicate", function()
    assert_eq(inbox.is_duplicate("msg-new-1", "test"), false)
end)

test("nil message_id returns false", function()
    assert_eq(inbox.is_duplicate(nil, "test"), false)
end)

test("empty message_id returns false", function()
    assert_eq(inbox.is_duplicate("", "test"), false)
end)

-- ── mark_processed ───────────────────────────────────────────────────

test("mark_processed then is_duplicate returns true", function()
    inbox.mark_processed("msg-mark-1", "test")
    assert_eq(inbox.is_duplicate("msg-mark-1", "test"), true)
end)

test("different source is not duplicate", function()
    inbox.mark_processed("msg-src-1", "source-A")
    assert_eq(inbox.is_duplicate("msg-src-1", "source-B"), false)
end)

-- ── check_and_mark ───────────────────────────────────────────────────

test("check_and_mark returns false for new message", function()
    local dup = inbox.check_and_mark("msg-cam-1", "test")
    assert_eq(dup, false, "first call should not be duplicate")
end)

test("check_and_mark returns true for second call", function()
    inbox.check_and_mark("msg-cam-2", "test")
    local dup = inbox.check_and_mark("msg-cam-2", "test")
    assert_eq(dup, true, "second call should be duplicate")
end)

-- ── default source ───────────────────────────────────────────────────

test("default source works", function()
    inbox.mark_processed("msg-default-1")
    assert_eq(inbox.is_duplicate("msg-default-1"), false)
    assert_eq(inbox.is_duplicate("msg-default-1", "default"), true)
end)

-- ── cleanup ─────────────────────────────────────────────────────────

test("cleanup returns count", function()
    local count = inbox.cleanup()
    assert(type(count) == "number", "expected number")
end)

return {pass = pass, fail = fail}
